app/src/main/java/org/openobservatory/ooniprobe/common/MKCollectorResubmitTask.java

```diff
@@ -0,0 +1,98 @@
+package org.openobservatory.ooniprobe.common;
+
+import android.content.Context;
+import android.util.Log;
+import android.view.WindowManager;
+
+import com.raizlabs.android.dbflow.sql.language.OperatorGroup;
+import com.raizlabs.android.dbflow.sql.language.SQLOperator;
+import com.raizlabs.android.dbflow.sql.language.SQLite;
+
+import org.apache.commons.io.FileUtils;
+import org.openobservatory.ooniprobe.R;
+import org.openobservatory.ooniprobe.model.database.Measurement;
+import org.openobservatory.ooniprobe.model.database.Measurement_Table;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import androidx.appcompat.app.AppCompatActivity;
+import io.ooni.mk.MKCollectorResubmitResults;
+import io.ooni.mk.MKCollectorResubmitSettings;
+import localhost.toolkit.os.NetworkProgressAsyncTask;
+
+public class MKCollectorResubmitTask<A extends AppCompatActivity> extends NetworkProgressAsyncTask<A, Integer, Void> {
+	/**
+	 * Use this class to resubmit a measurement, use result_id and measurement_id to filter list of value
+	 * {@code new MKCollectorResubmitTask(activity).execute(@Nullable result_id, @Nullable measurement_id);}
+	 *
+	 * @param activity from which this task are executed
+	 */
+	public MKCollectorResubmitTask(A activity) {
+		super(activity, true, false);
+	}
+
+	private static void perform(Context c, Measurement m) throws IOException {
+		String input = FileUtils.readFileToString(Measurement.getEntryFile(c, m.id, m.test_name), StandardCharsets.UTF_8);
+		MKCollectorResubmitSettings settings = new MKCollectorResubmitSettings();
+		settings.setTimeout(14);
+		settings.setCABundlePath(c.getCacheDir() + "/" + Application.CA_BUNDLE);
+		settings.setSerializedMeasurement(input);
+		MKCollectorResubmitResults results = settings.perform();
+		if (results.isGood()) {
+			String output = results.getUpdatedSerializedMeasurement();
+			FileUtils.writeStringToFile(Measurement.getEntryFile(c, m.id, m.test_name), output, StandardCharsets.UTF_8);
+			m.report_id = results.getUpdatedReportID();
+			m.is_uploaded = true;
+			m.is_upload_failed = false;
+			m.save();
+		} else {
+			Log.w(MKCollectorResubmitSettings.class.getSimpleName(), results.getLogs());
+			// TODO decide what to do with logs (append on log file?)
+		}
+	}
+
+	@Override protected void onPreExecute() {
+		super.onPreExecute();
+		A activity = getActivity();
+		if (activity != null)
+			activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
+	}
+
+	@Override protected Void doInBackground(Integer... params) {
+		if (params.length != 2)
+			throw new IllegalArgumentException("MKCollectorResubmitTask requires 2 nullable params: result_id, measurement_id");
+		ArrayList<SQLOperator> where = new ArrayList<>();
+		if (params[0] != null)
+			where.add(Measurement_Table.result_id.eq(params[0]));
+		if (params[1] != null)
```

Comment:
Will this ever be called with a `params` array which has two null entries? If not we should maybe explicitly check this at the beginning and throw an expection if that is the case.

I think this would also allow us to get rid of this `ArrayList` for the where clause and reduce complexity.

Reply:
it will be called in all combination

Reply:
> it will be called in all combination

You mean that it's possible for `param[0]` to be `null`? In which case will that happen?

Maybe what you can do if that is the case is define a `result_id` variable and then use the syntax sugar of your ORM to form the query directly without passing via an array.

Example:

```
		if (params.length != 2)
			throw new IllegalArgumentException("MKCollectorResubmitTask requires 2 nullable params: result_id, measurement_id");
		String result_id;
		if (params[0] != null) {
                    result_id = params[0]
                 } else if (params[1] != null) {
                     result_id = params[1]
                } else {
                    // XXX is this even possible? should we maybe throw an exception?
                }
		where.add(.or().or());
		where.add();
		List<Measurement> measurements = SQLite.select().from(Measurement.class)
                  .where(Measurement_Table.is_failed.eq(false))
                 // We check on both the report_id and is_uploaded as we may have some unuploaded measurements which are marked as is_uploaded = true, but we always know that those with report_id set to null are not uploaded
                  .and(
                      OperatorGroup.clause().or(Measurement_Table.is_uploaded.eq(false))
                      .or(Measurement_Table.report_id.isNull()
                  )
		.and(Measurement_Table.result_id.eq(result_id))
                 .queryList();

```

Reply:
params[0] are result_id
params[1] are measurement_id

Null null means get all measurements

Reply:
Ok I understand now. Thanks for explaining.

How about then doing something like:
```
	 msmQuery = SQLite.select().from(Measurement.class)
                  .where(Measurement_Table.is_failed.eq(false))
                 // We check on both the report_id and is_uploaded as we 
                // may have some unuploaded measurements which are marked 
               // as is_uploaded = true, but we always know that those with 
               // report_id set to null are not uploaded
                  .and(
                      OperatorGroup.clause().or(Measurement_Table.is_uploaded.eq(false))
                      .or(Measurement_Table.report_id.isNull()
                  )
	if (params[0] != null) {
                    msmQuery.and(Measurement_Table.result_id.eq(result_id))
                 } else if (params[1] != null) {
                    msmQuery.and(Measurement_Table.result_id.eq(params[0]))
               }
             List<Measurement> measurements = msmQuery.queryList()
```

Reply:
done, but i don't see much difference with arraylist